source/common/http/http2/metadata_encoder.cc

```diff
@@ -79,14 +84,21 @@ bool MetadataEncoder::createHeaderBlockUsingNghttp2(const MetadataMap& metadata_
 
 bool MetadataEncoder::hasNextFrame() { return payload_.length() > 0; }
 
-uint64_t MetadataEncoder::packNextFramePayload(uint8_t* buf, const size_t len) {
-  ASSERT(!payload_size_queue_.empty());
+ssize_t MetadataEncoder::packNextFramePayload(uint8_t* buf, const size_t len) {
+  if (payload_size_queue_.empty()) {
+    ENVOY_LOG(error, "No payload remaining to pack into a METADATA frame.");
+    return -1;
+  }
   const uint64_t current_payload_size =
       std::min(METADATA_MAX_PAYLOAD_SIZE, payload_size_queue_.front());
 
   // nghttp2 guarantees len is at least 16KiB. If the check fails, please verify
   // NGHTTP2_MAX_PAYLOADLEN is consistent with METADATA_MAX_PAYLOAD_SIZE.
-  ASSERT(len >= current_payload_size);
+  if (len < current_payload_size) {
```

Comment:
It can't, for two reasons: NGHTTP2_MAX_PAYLOADLEN isn't in the nghttp2 public API, and the static_assert wouldn't guarantee that nghttp2 respected that macro.